### ISSUE TITLE
Link a workflow to a form

### ISSUE THREAD

**Issue by sfmskywalker** [OrchardCMS/OrchardCore#1820]
Submitting the form will then trigger that specific workflow. The form will then submit to a specific FormWorkflowController that will take care of

1. Antiforgery token validation (if enabled on the form)
2. Form field validation
3. Bind form fields to model state

**Comment by sfmskywalker** [comment]
See also: #1819

**Comment by agriffard** [comment]
Isn't it already done with the HttpRequest event?

**Comment by sfmskywalker** [comment]
Nah, this is about visually connecting a form widget with a workflow, e.g. by selecting a workflow from a drop down. The form essentially becomes the “code behind” of the form.

**Comment by deanmarcussen** [comment]
@sfmskywalker I appear to be working on something to do with this, and found some old code of yours, that is disabled/ unused - `SelectWorkflowTypeViewComponent`

The approach I am taking is somewhat different maybe to what you had intended, so worth gathering some info.

I'm working on a Secrets module, which allows an abstraction of things like Passwords, RSA keys - you remember our conversations about how to manage an email password through configuration.

The feature is mostly about being able to transfer data between sites which do not share the same encryption keys, because of course the email password is encrypted for that site only, so doesn't transfer properly.

The thought is to have another `HttpWorkflowSecret` for the `HttpWorkflowEvent` which contains/generates the token details, and becomes transportable across sites via some RSA encryption.

That secret is selectable / generatable from the `HttpWorkflowController`

Then on the `FormPart` an dropdown allows you to optional select the `HttpWorkflowSecret` instead of the url (you can still use a url if you want to).

When it generates it will retrieve the correct endpoint from the service.

Early demo here https://www.youtube.com/watch?v=qPCBgHQYz1g

Any thoughts appreciated

**Comment by sfmskywalker** [comment]
The Secrets module makes a lot of sense, nice work!

Regarding the **Select Workflow Type** view component, its purpose is not to trigger a workflow using HTTP, but to actually invoke the selected workflow, regardless of its starting activity being an event or not. For example, such a selected workflow might start immediately by storing the submitted form values in some database. No HTTP events involved.
But that's just one use case (which I never got to finish).

So yeah, in addition to being able to enter a URL, it would be quite nifty to be able to select a preconfigured HTTP endpoint in the form of an `HttpWorkflowSecret`.

And if you are already allowing the user to either specify a URL (free text form) or an HttpWorkflowSecret (finite set of options), a third option might exist as well allowing the user to select a workflow to invoke directly as I just outlined.

**Comment by sfmskywalker** [comment]
Regarding the discussion about **secret binding** vs **secret**, it's an interesting one. One can bind a configuration source against some options object, but in Orchard, there's of course the `IConfiguration` API built from various sources such as appsettings.json and ENV, but also Orchard Settings which can be managed at runtime by a user. In any case, this is all databinding a IConfiguration against some strongly-typed class, so I think I get the term "binding" in this context.

But from a user's perspective, being on the **Edit Secret Bindings** screen, it seems to me that what a user is doing there, as far as they are concerned, is adding a secret to the selected store. The fact that these secrets are bound somewhere against something is just an implementation detail as far as the user is concerned?

**Comment by sfmskywalker** [comment]
> And if you are already allowing the user to either specify a URL (free text form) or an HttpWorkflowSecret (finite set of options), a third option might exist as well allowing the user to select a workflow to invoke directly as I just outlined.

On second thought, after watching the video some more, I just realized that the view component you are working on is really just a secret editor allowing you to either select an existing secret or create a new secret on demand. So forget what I said earlier.

Being able to select an arbitrary workflow to execute upon form submission should be considered out of scope and be revisited when the need arises.

**Comment by deanmarcussen** [comment]
We love a good naming debate ;)

So I think I ended up agreeing better to just use `Secret` in the UI, as you're right, that is what you're editing. 

I think I kept it in the demo, because I wanted to demo / easily explain that it was both a binding arrangement, and possibly creating a secret.

e.g. the secrets that come from an `IConfiguration` secret store, are readonly, so they're already there in `IShellConfiguration`, waiting for someone to bind them.

Imagine for example on a Saas site, you say, email is a free service, we store it in a global configuration for you, to use the free service, create a secret called `free_email_password`, and select the `Configuration Store` and it'll all start working.

But from a user point of view, they don't need to know that underneath the hood it's a Secret Binding, for them it's just a Secret...

Cool, I'll carry on with what has become a `HttpRequestEventSecret` - in and working, just seeing how the ui will play out now for it, and we can come back to this form submission another time.
